File: examples/flutter/lib/mijauth.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:crypto/crypto.dart';

class MijAuth {
  static const String cipher = 'aes-256-gcm';
  static const int keyLength = 32;
  static const int ivLength = 12;
  static const int tagLength = 16;
  static const int version = 1;

  static const String libraryVersion = '0.3.0';

  static final AesGcm _algorithm = AesGcm.with256bits();

  static String generateUserKey() {
    final bytes = _randomBytes(keyLength);
    return base64Encode(bytes);
  }

  static String generateToken() {
    return _bytesToHex(_randomBytes(32));
  }

  static Future<Map<String, String>> createAuthFile(
    String userId,
    String userKeyBase64, {
    String? deviceHash,
    String? deviceHashV2,
  }) async {
    final token = generateToken();

    final payload = <String, dynamic>{
      'user_id': userId,
      'token': token,
      'created_at': DateTime.now().toUtc().toIso8601String(),
      'device_hash': deviceHash,
      'device_hash_v2': deviceHashV2,
      'version': version,
    };

    final jsonPayload = jsonEncode(payload);
    final encryptedContent = await encrypt(jsonPayload, userKeyBase64);

    return {
      'file_content': encryptedContent,
      'token': token,
    };
  }

  static Future<Map<String, dynamic>?> verifyAuthFile(
    String fileContent,
    String userKeyBase64,
  ) async {
    try {
      final decrypted = await decrypt(fileContent, userKeyBase64);
      if (decrypted == null) {
        return null;
      }

      final payload = jsonDecode(decrypted) as Map<String, dynamic>;

      if (!payload.containsKey('user_id') ||
          !payload.containsKey('token') ||
          !payload.containsKey('version')) {
        return null;
      }

      return payload;
    } catch (_) {
      return null;
    }
  }

  static Future<bool> verifyAuthFileWithToken(
    String fileContent,
    String userKeyBase64,
    String expectedToken,
    String expectedUserId,
  ) async {
    final payload = await verifyAuthFile(fileContent, userKeyBase64);
    if (payload == null) {
      return false;
    }

    return _timingSafeEquals(expectedToken, payload['token'] as String) &&
        _timingSafeEquals(expectedUserId, payload['user_id'] as String);
  }

  static Future<bool> verifyAuthFileWithTokenAndDevice(
    String fileContent,
    String userKeyBase64,
    String expectedToken,
    String expectedUserId, {
    String? expectedDeviceHash,
    String? expectedDeviceHashV2,
  }) async {
    final payload = await verifyAuthFile(fileContent, userKeyBase64);
    if (payload == null) {
      return false;
    }

    if (!_timingSafeEquals(expectedToken, payload['token'] as String) ||
        !_timingSafeEquals(expectedUserId, payload['user_id'] as String)) {
      return false;
    }

    if (expectedDeviceHash != null) {
      if (!payload.containsKey('device_hash') ||
          !_timingSafeEquals(expectedDeviceHash, payload['device_hash'] as String)) {
        return false;
      }
    }

    if (expectedDeviceHashV2 != null) {
      if (!payload.containsKey('device_hash_v2') ||
          !_timingSafeEquals(expectedDeviceHashV2, payload['device_hash_v2'] as String)) {
        return false;
      }
    }

    return true;
  }

  static Future<Map<String, String>> regenerateAuthFile(
    String userId,
    String userKeyBase64, {
    String? deviceHash,
    String? deviceHashV2,
  }) {
    return createAuthFile(
      userId,
      userKeyBase64,
      deviceHash: deviceHash,
      deviceHashV2: deviceHashV2,
    );
  }

  static String generateDeviceHash(
    String userAgent,
    String acceptLanguage, {
    Map<String, dynamic> additionalData = const {},
  }) {
    final data = <String, dynamic>{
      'user_agent': userAgent,
      'accept_language': acceptLanguage,
      ...additionalData,
    };

    return sha256.convert(utf8.encode(jsonEncode(data))).toString();
  }

  static String generateDeviceHashV2(Map<String, dynamic> context) {
    final normalized = _normalizeFingerprintContext(context);
    return sha256.convert(utf8.encode(jsonEncode(normalized))).toString();
  }

  static Future<String> encrypt(String plaintext, String keyBase64) async {
    final keyBytes = base64Decode(keyBase64);
    if (keyBytes.length != keyLength) {
      throw ArgumentError('Invalid encryption key');
    }

    final secretKey = SecretKey(keyBytes);
    final nonce = _randomBytes(ivLength);

    final secretBox = await _algorithm.encrypt(
      utf8.encode(plaintext),
      secretKey: secretKey,
      nonce: nonce,
    );

    final combined = Uint8List(ivLength + tagLength + secretBox.cipherText.length);
    combined.setRange(0, ivLength, secretBox.nonce);
    combined.setRange(ivLength, ivLength + tagLength, secretBox.mac.bytes);
    combined.setRange(ivLength + tagLength, combined.length, secretBox.cipherText);

    return base64Encode(combined);
  }

  static Future<String?> decrypt(String encryptedBase64, String keyBase64) async {
    final keyBytes = base64Decode(keyBase64);
    if (keyBytes.length != keyLength) {
      return null;
    }

    final combined = base64Decode(encryptedBase64);
    if (combined.length < ivLength + tagLength) {
      return null;
    }

    final nonce = combined.sublist(0, ivLength);
    final macBytes = combined.sublist(ivLength, ivLength + tagLength);
    final cipherText = combined.sublist(ivLength + tagLength);

    final secretBox = SecretBox(
      cipherText,
      nonce: nonce,
      mac: Mac(macBytes),
    );

    try {
      final clearText = await _algorithm.decrypt(
        secretBox,
        secretKey: SecretKey(keyBytes),
      );
      return utf8.decode(clearText);
    } catch (_) {
      return null;
    }
  }

  static List<int> _randomBytes(int length) {
    final random = Random.secure();
    return List<int>.generate(length, (_) => random.nextInt(256));
  }

  static String _bytesToHex(List<int> bytes) {
    final buffer = StringBuffer();
    for (final b in bytes) {
      buffer.write(b.toRadixString(16).padLeft(2, '0'));
    }
    return buffer.toString();
  }

  static bool _timingSafeEquals(String a, String b) {
    final aBytes = utf8.encode(a);
    final bBytes = utf8.encode(b);

    if (aBytes.length != bBytes.length) {
      return false;
    }

    var result = 0;
    for (var i = 0; i < aBytes.length; i++) {
      result |= aBytes[i] ^ bBytes[i];
    }
    return result == 0;
  }

  static bool timingSafeEquals(String a, String b) {
    return _timingSafeEquals(a, b);
  }

  static Map<String, dynamic> _normalizeFingerprintContext(
    Map<String, dynamic> context,
  ) {
    final normalized = <String, dynamic>{};

    for (final entry in context.entries) {
      final value = entry.value;
      if (value == null) {
        continue;
      }
      if (value is Map<String, dynamic>) {
        normalized[entry.key] = _normalizeFingerprintContext(value);
      } else {
        normalized[entry.key] = value;
      }
    }

    final sortedKeys = normalized.keys.toList()..sort();
    return {for (final key in sortedKeys) key: normalized[key]};
  }
}

class DemoUser {
  DemoUser({
    required this.id,
    required this.email,
    required this.passwordHash,
    required this.encryptionKey,
    required this.authToken,
  });

  final String id;
  final String email;
  final String passwordHash;
  final String encryptionKey;
  String authToken;
}

class DemoUserDatabase {
  final Map<String, DemoUser> _users = {};

  DemoUser? getUserByEmail(String email) {
    for (final user in _users.values) {
      if (user.email == email) {
        return user;
      }
    }
    return null;
  }

  DemoUser? getUser(String userId) {
    return _users[userId];
  }

  Future<Map<String, String>> createUser(
    String userId,
    String email,
    String password,
  ) async {
    final userKey = MijAuth.generateUserKey();
    final authResult = await MijAuth.createAuthFile(userId, userKey);

    final passwordHash = sha256
        .convert(utf8.encode(password + 'salt_' + userId))
        .toString();

    final user = DemoUser(
      id: userId,
      email: email,
      passwordHash: passwordHash,
      encryptionKey: userKey,
      authToken: authResult['token']!,
    );

    _users[userId] = user;

    return {
      'auth_file': authResult['file_content']!,
      'token': authResult['token']!,
    };
  }

  bool verifyPassword(DemoUser user, String password) {
    final hash = sha256
        .convert(utf8.encode(password + 'salt_' + user.id))
        .toString();
    return MijAuth.timingSafeEquals(hash, user.passwordHash);
  }

  void updateAuthToken(String userId, String newToken) {
    final user = _users[userId];
    if (user != null) {
      user.authToken = newToken;
    }
  }
}

File: examples/kotlin/MijAuth.kt
import java.io.File
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Instant
import java.time.Duration
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.math.max

object MijAuth {
    private const val KEY_LENGTH = 32
    private const val IV_LENGTH = 12
    private const val TAG_LENGTH_BITS = 128
    private const val VERSION = 1
    private const val DEFAULT_AUTH_FILE_TTL_SECONDS = 30L * 24L * 60L * 60L

    data class AuthPayload(
        val user_id: String,
        val token: String,
        val created_at: String,
        val device_hash: String? = null,
        val device_hash_v2: String? = null,
        val version: Int = VERSION
    )

    data class User(
        val id: String,
        val email: String,
        val password_hash: String,
        val encryption_key: String,
        var auth_token: String,
        val totp_secret: String,
        val created_at: String
    )

    fun generateUserKey(): String {
        val key = ByteArray(KEY_LENGTH)
        SecureRandom().nextBytes(key)
        return Base64.getEncoder().encodeToString(key)
    }

    fun generateToken(): String {
        val token = ByteArray(32)
        SecureRandom().nextBytes(token)
        return token.joinToString("") { "%02x".format(it) }
    }

    fun createAuthFile(
        userId: String,
        userKeyBase64: String,
        deviceHash: String? = null,
        deviceHashV2: String? = null
    ): Pair<String, String> {
        val token = generateToken()
        val payload = AuthPayload(
            user_id = userId,
            token = token,
            created_at = Instant.now().toString(),
            device_hash = deviceHash,
            device_hash_v2 = deviceHashV2
        )

        val jsonPayload = toJson(payload)
        return encrypt(jsonPayload, userKeyBase64) to token
    }

    fun verifyAuthFile(
        fileContent: String,
        userKeyBase64: String,
        maxAgeSeconds: Long? = DEFAULT_AUTH_FILE_TTL_SECONDS
    ): AuthPayload? {
        val decrypted = decrypt(fileContent, userKeyBase64) ?: return null
        val payload = parsePayload(decrypted) ?: return null

        if (payload.user_id.isBlank() || payload.token.isBlank()) return null

        if (maxAgeSeconds != null) {
            val createdAt = runCatching { Instant.parse(payload.created_at) }.getOrNull() ?: return null
            val now = Instant.now()
            if (createdAt.isAfter(now)) return null
            if (Duration.between(createdAt, now).seconds > maxAgeSeconds) return null
        }

        return payload
    }

    fun verifyAuthFileWithToken(
        fileContent: String,
        userKeyBase64: String,
        expectedToken: String,
        expectedUserId: String,
        maxAgeSeconds: Long? = DEFAULT_AUTH_FILE_TTL_SECONDS
    ): Boolean {
        val payload = verifyAuthFile(fileContent, userKeyBase64, maxAgeSeconds) ?: return false
        return fixedTimeEquals(expectedToken, payload.token) && fixedTimeEquals(expectedUserId, payload.user_id)
    }

    fun regenerateAuthFile(
        userId: String,
        userKeyBase64: String,
        deviceHash: String? = null,
        deviceHashV2: String? = null
    ): Pair<String, String> = createAuthFile(userId, userKeyBase64, deviceHash, deviceHashV2)

    fun generateTotpSecret(length: Int = 32): String {
        require(length >= 16) { "Sekret TOTP musi mieć co najmniej 16 znaków" }
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        val random = ByteArray(length)
        SecureRandom().nextBytes(random)
        return buildString(length) {
            random.forEach { append(alphabet[(it.toInt() and 0xff) % alphabet.length]) }
        }
    }

    fun getTotpProvisioningUri(
        accountName: String,
        issuer: String,
        secret: String,
        digits: Int = 6,
        period: Int = 30
    ): String {
        val label = java.net.URLEncoder.encode("$issuer:$accountName", Charsets.UTF_8)
        val query = listOf(
            "secret=${java.net.URLEncoder.encode(secret.uppercase(), Charsets.UTF_8)}",
            "issuer=${java.net.URLEncoder.encode(issuer, Charsets.UTF_8)}",
            "algorithm=SHA1",
            "digits=$digits",
            "period=$period"
        ).joinToString("&")

        return "otpauth://totp/$label?$query"
    }

    fun generateTotpCode(
        secret: String,
        timestampSeconds: Long = Instant.now().epochSecond,
        period: Int = 30,
        digits: Int = 6
    ): String {
        val counter = timestampSeconds / period
        return generateHotpCode(secret, counter, digits)
    }

    fun verifyTotp(
        secret: String,
        code: String,
        discrepancy: Int = 1,
        timestampSeconds: Long = Instant.now().epochSecond,
        period: Int = 30,
        digits: Int = 6
    ): Boolean {
        val normalizedCode = code.replace("\\s+".toRegex(), "")
        if (!normalizedCode.matches(Regex("^\\\\d{$digits}$"))) return false

        val baseCounter = timestampSeconds / period
        for (offset in -discrepancy..discrepancy) {
            val candidate = generateHotpCode(secret, baseCounter + offset, digits)
            if (fixedTimeEquals(candidate, normalizedCode)) return true
        }

        return false
    }

    private fun generateHotpCode(secret: String, counter: Long, digits: Int): String {
        if (counter < 0) return "0".repeat(digits)

        val key = decodeBase32(secret)
        require(key.isNotEmpty()) { "Nieprawidłowy sekret TOTP" }

        val counterBytes = ByteArray(8)
        var value = counter
        for (i in 7 downTo 0) {
            counterBytes[i] = (value and 0xff).toByte()
            value = value shr 8
        }

        val mac = Mac.getInstance("HmacSHA1")
        mac.init(SecretKeySpec(key, "HmacSHA1"))
        val hash = mac.doFinal(counterBytes)
        val offset = hash.last().toInt() and 0x0f
        val binary =
            ((hash[offset].toInt() and 0x7f) shl 24) or
            ((hash[offset + 1].toInt() and 0xff) shl 16) or
            ((hash[offset + 2].toInt() and 0xff) shl 8) or
            (hash[offset + 3].toInt() and 0xff)

        val mod = Math.pow(10.0, digits.toDouble()).toInt()
        return (binary % mod).toString().padStart(digits, '0')
    }

    private fun decodeBase32(secret: String): ByteArray {
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        val normalized = secret.replace("=", "").replace("\\s+".toRegex(), "").uppercase()
        val bits = StringBuilder()

        for (ch in normalized) {
            val idx = alphabet.indexOf(ch)
            if (idx < 0) return ByteArray(0)
            bits.append(idx.toString(2).padStart(5, '0'))
        }

        val bytes = ArrayList<Byte>()
        var i = 0
        while (i + 8 <= bits.length) {
            val byteValue = bits.substring(i, i + 8).toInt(2)
            bytes.add(byteValue.toByte())
            i += 8
        }

        return bytes.toByteArray()
    }

    private fun encrypt(plaintext: String, keyBase64: String): String {
        val key = Base64.getDecoder().decode(keyBase64)
        val iv = ByteArray(IV_LENGTH)
        SecureRandom().nextBytes(iv)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val gcm = GCMParameterSpec(TAG_LENGTH_BITS, iv)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), gcm)
        val ciphertextWithTag = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))

        val combined = iv + ciphertextWithTag
        return Base64.getEncoder().encodeToString(combined)
    }

    private fun decrypt(encryptedBase64: String, keyBase64: String): String? {
        return try {
            val key = Base64.getDecoder().decode(keyBase64)
            val combined = Base64.getDecoder().decode(encryptedBase64)
            if (combined.size < IV_LENGTH + 16) return null

            val iv = combined.copyOfRange(0, IV_LENGTH)
            val ciphertextWithTag = combined.copyOfRange(IV_LENGTH, combined.size)

            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            val gcm = GCMParameterSpec(TAG_LENGTH_BITS, iv)
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), gcm)
            val plaintext = cipher.doFinal(ciphertextWithTag)
            plaintext.toString(Charsets.UTF_8)
        } catch (_: Exception) {
            null
        }
    }

    fun generateDeviceHash(userAgent: String = "", acceptLanguage: String = "", deviceId: String = ""): String {
        val data = "{\"user_agent\":\"$userAgent\",\"accept_language\":\"$acceptLanguage\",\"device_id\":\"$deviceId\"}"
        return sha256Hex(data)
    }

    private fun toJson(payload: AuthPayload): String {
        fun n(v: String?) = if (v == null) "null" else "\"${v.replace("\"", "\\\"")}\""
        return "{" +
            "\"user_id\":${n(payload.user_id)}," +
            "\"token\":${n(payload.token)}," +
            "\"created_at\":${n(payload.created_at)}," +
            "\"device_hash\":${n(payload.device_hash)}," +
            "\"device_hash_v2\":${n(payload.device_hash_v2)}," +
            "\"version\":${payload.version}" +
            "}"
    }

    private fun parsePayload(json: String): AuthPayload? {
        val regex = Regex("\"(user_id|token|created_at|device_hash|device_hash_v2|version)\"\\s*:\\s*(null|\"[^\"]*\"|\\d+)")
        val values = mutableMapOf<String, String?>()
        regex.findAll(json).forEach { match ->
            val key = match.groupValues[1]
            val raw = match.groupValues[2]
            values[key] = if (raw == "null") null else raw.trim('"')
        }

        val userId = values["user_id"] ?: return null
        val token = values["token"] ?: return null
        val createdAt = values["created_at"] ?: return null
        val version = values["version"]?.toIntOrNull() ?: VERSION

        return AuthPayload(
            user_id = userId,
            token = token,
            created_at = createdAt,
            device_hash = values["device_hash"],
            device_hash_v2 = values["device_hash_v2"],
            version = version
        )
    }

    private fun fixedTimeEquals(a: String, b: String): Boolean {
        val aa = a.toByteArray(Charsets.UTF_8)
        val bb = b.toByteArray(Charsets.UTF_8)
        if (aa.size != bb.size) return false

        var diff = 0
        for (i in aa.indices) diff = diff or (aa[i].toInt() xor bb[i].toInt())
        return diff == 0
    }

    private fun sha256Hex(value: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(value.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}

class UserDatabase(private val storageFile: String = "users.json") {
    private val users = mutableMapOf<String, MijAuth.User>()

    init {
        load()
    }

    fun createUser(userId: String, email: String, password: String): Pair<MijAuth.User, String> {
        val userKey = MijAuth.generateUserKey()
        val (authFile, token) = MijAuth.createAuthFile(userId, userKey)
        val totpSecret = MijAuth.generateTotpSecret()
        val passwordHash = sha256Hex(password + "salt_" + userId)

        val user = MijAuth.User(
            id = userId,
            email = email,
            password_hash = passwordHash,
            encryption_key = userKey,
            auth_token = token,
            totp_secret = totpSecret,
            created_at = Instant.now().toString()
        )

        users[userId] = user
        save()

        return user to authFile
    }

    fun getUser(userId: String): MijAuth.User? = users[userId]

    fun getUserByEmail(email: String): MijAuth.User? = users.values.firstOrNull { it.email == email }

    fun verifyPassword(user: MijAuth.User, password: String): Boolean {
        return sha256Hex(password + "salt_" + user.id) == user.password_hash
    }

    fun updateAuthToken(userId: String, newToken: String) {
        val existing = users[userId] ?: return
        users[userId] = existing.copy(auth_token = newToken)
        save()
    }

    fun deleteStorage() {
        File(storageFile).delete()
    }

    private fun load() {
        val file = File(storageFile)
        if (!file.exists()) return

        val text = file.readText()
        val objectRegex = Regex("\\{[^{}]*}")
        objectRegex.findAll(text).forEach { match ->
            val user = parseUser(match.value)
            if (user != null) users[user.id] = user
        }
    }

    private fun save() {
        val serialized = users.entries.joinToString(",\n", prefix = "{\n", postfix = "\n}") { (id, u) ->
            "  \"$id\": {\"id\":\"${u.id}\",\"email\":\"${u.email}\",\"password_hash\":\"${u.password_hash}\",\"encryption_key\":\"${u.encryption_key}\",\"auth_token\":\"${u.auth_token}\",\"totp_secret\":\"${u.totp_secret}\",\"created_at\":\"${u.created_at}\"}"
        }
        File(storageFile).writeText(serialized)
    }

    private fun parseUser(json: String): MijAuth.User? {
        val regex = Regex("\"(id|email|password_hash|encryption_key|auth_token|totp_secret|created_at)\"\\s*:\\s*\"([^\"]*)\"")
        val values = mutableMapOf<String, String>()
        regex.findAll(json).forEach { values[it.groupValues[1]] = it.groupValues[2] }

        val id = values["id"] ?: return null
        return MijAuth.User(
            id = id,
            email = values["email"] ?: return null,
            password_hash = values["password_hash"] ?: return null,
            encryption_key = values["encryption_key"] ?: return null,
            auth_token = values["auth_token"] ?: return null,
            totp_secret = values["totp_secret"] ?: return null,
            created_at = values["created_at"] ?: return null
        )
    }

    private fun sha256Hex(value: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(value.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}
